src/Microsoft.AspNetCore.Blazor.Browser/Interop/JavaScriptInvoke.cs

```diff
@@ -0,0 +1,253 @@
+            return (string args) => method.Invoke(null, argumentDeserializer(args));
+        }
+
+        private static Func<string, object[]> GetOrCreateArgumentDeserializer(MethodOptions options)
+        {
+            var info = options.GetMethodOrThrow();
+            var argsClass = ArgumentList.GetArgumentClass(info.GetParameters().Select(p => p.ParameterType).ToArray());
+            var deserializeMethod = ArgumentList.GetDeserializer(argsClass);
+            var toParameterListMethod = argsClass.GetMethods(BindingFlags.Instance | BindingFlags.Public)
+                .Where(m => string.Equals(nameof(ArgumentList.ToParameterList), m.Name))
+                .Single();
+
+            return Deserialize;
+
+            object[] Deserialize(string arguments)
+            {
+                var argsInstance = deserializeMethod(arguments);
+                return argsInstance.ToParameterList();
+            }
+        }
+
+        /// <summary>
+        /// TODO
+        /// </summary>
+        /// <param name="identifier"></param>
+        /// <param name="args"></param>
+        /// <returns></returns>
+        public static Task InvokeJavaScriptVoidFunctionAsync(string identifier, params object[] args)
+        {
+            var tcs = new TaskCompletionSource<int>();
+            var argsJson = args.Select(JsonUtil.Serialize);
+            var successId = Guid.NewGuid().ToString();
+            var failureId = Guid.NewGuid().ToString();
+            var asyncProtocol = JsonUtil.Serialize(new
+            {
+                Success = successId,
+                Failure = failureId,
+                Function = new MethodOptions
+                {
+                    Type = new TypeInstance
+                    {
+                        Assembly = typeof(JavaScriptInvoke).Assembly.FullName,
+                        TypeName = typeof(JavaScriptInvoke).FullName
+                    },
+                    Method = new MethodInstance
+                    {
+                        Name = nameof(JavaScriptInvoke.InvokeTaskCallback),
+                        ParameterTypes = new[]
+                        {
+                            new TypeInstance
+                            {
+                                Assembly = typeof(string).Assembly.FullName,
+                                TypeName = typeof(string).FullName
+                            },
+                            new TypeInstance
+                            {
+                                Assembly = typeof(string).Assembly.FullName,
+                                TypeName = typeof(string).FullName
+                            }
+                        }
+                    }
+                }
+            });
+
+            TrackedReference.Track(successId, (new Action(() =>
+            {
+                tcs.SetResult(0);
+            })));
+
+            TrackedReference.Track(failureId, (new Action<string>(r =>
```

Comment:
Instead of tracking these two actions separately, could we track the `TaskCompletionSource` itself? We're only going to use one of the two continuations (success or failure) so one of the two tracked items would be redundant. Not a massive deal but looks like it would save two heap allocations per interop call.


That would save one heap allocation per call, since we're only going to use one of the two continuations (success or failure, not both).

Reply:
I've already changed the implementation to require only 1 callback and have it handle both scenarios as part of the exceptions work.